README.md
USAGE
-----

    $ ./insub.py <<<'"hi"|figlet|cow'
     ___________
    /  _     _  \
    | | |__ (_) |
    | | '_ \| | |
    | | | | | | |
    | |_| |_|_| |
    \           /
     -----------
            \   ^__^
             \  (oo)_______
                (__)       )/\
                    ||----w |
                    ||     ||

... it does a bit more than that, but that is the gist of it. For a list
of all the filters, run with -h


IRSSI
-----

irssi/insub.pl is a thin perl wrapper around the python script, which is
a total hack, but works. I am told irssi supports python embedding
natively, but it doesn't seem to in Ubuntu's repository, which is what I
use, so whatever. Use it like any other irssi module:

    $ cp -av irssi/insub.pl ~/.irssi/scripts/
    [irssi] /script load insub


LEGACY
------

The original version (of both names) can be found in the old/ directory.
It mostly works the same, except does not support unicode in any form, and
is written in terrible perl.


old/ChangeLog
50.0: Wed Nov 16 01:43:20 2016
    * Add half-assed slack support. Use an "api test token" so it can spoof
      your user name. https://api.slack.com/docs/oauth-test-tokens
    * Fix the startup splash message to not lie about how to get help. That alone merits a bump to 50.x

16.0: Thu Jul  4 16:19:33 2013
    * Some unicode support thanks to @obsdler
    * Updated -spook words with PRISM list of watch-words (supposedly).

15.1: Sun Aug 19 05:30:51 2012
    * Separate value add cows from cowsay cows

15.0: Sat Aug 18 13:07:55 2012
    * Fix the fucking URL in /gv.
    * JESUS CHRIST I'M LIKE FIREFOX

14.3: Sat Aug 18 13:02:04 2012
    * Fix -update FOR REAL

14.2: Sat Aug 18 12:49:43 2012
    * Uh, fix -update ...

14.1: Sat Aug 18 12:42:10 2012
    * Fix for ubuntu's broken irssi

14.0: Wed Aug 15 16:17:30 2012
    * BAM! BSD license.
    * Fix broken links/emails
    * Added a fuckton of cows. MANY COWS.

13.0: Mon Sep  1 12:51:39 2008
    * fixed missing -3 (random) style for rainbow

12.0: Sat Jul 14 14:33:09 2007
    * new flag -rain lets you specify a custom rainbow colormap. for
      example -rain rrwwBB will cause the rainbow to be 3 bars of red, 3
      bars of white, and 3 bars of bright blue
    * new flag -skew lets you set how the rainbow get skewed each line.
      default is 1 which causes the rainbow bars to appear diagonally
      bottom-left to top-right. -1 would cause them to reverse directions.
      higher skew values will create a sharper grade
    * add -chalkboard flag, graphic of your text being written by bart
    * remove annoying warning if you try to do too much figlet
    * allow 5ms throttling
    * fix bug with banner
    * switch to GPL license
    * autoflushing always on
    * a bunch of code cleanup
    * fix outdated contact info
    * remove unsightly tab from usage menu
    * completely redo the rainbow function
    * updated the usage hints
    * made help/colors/version/usage commands take a -

11.0: Tue May 29 23:05:40 2007

10.1: Wed Nov 22 14:50:12 2006
    * added silc client support

10.0: Sun May 29 01:06:10 2005
    * some fixes from zb

9.8: Thu Feb 17 03:19:08 2005
    * implement port of banner in perl, for hosts that don't have it

9.7: Wed Feb 16 10:03:29 2005
    * added bens banner wrappers
    * some backend code cleanup

9.6: Sat Dec  4 14:45:40 2004
    * fix critical bug in arg processing

9.5: Sat Dec  4 01:17:18 2004
    * moved -exec command into process(), so irssi can use it
    * fix bug that wouldn't let '0' be printed by itself
    * add -repeat flag ("-repeat 3 foo" becomes foofoofoo)
    * fix bug that would prevent a single character from displaying
    * retool -sine iterator a little better
    * minor code cleanup
    * make both -wave and -sine do the same thing

9.4: Sun Nov 28 20:02:02 2004
    * added sine-wave function (use -wave)
    * added zb's color optimization changes

9.3: Sun Nov 14 18:31:00 2004
    * added -popeye and -diagonal command
    * made -YES case-insensitive
    * moved order that -matrix is run in for consistency
    * cleaned up changelog
    * updated copyright

9.2: Tue Jul 20 21:42:04 2004
    * Fixed critical vulnerability which causes user to not be able to
      figlet a single character.

9.1: Sun Jul  4 01:15:43 2004
    * added -rotate, which flips a figlet 90 degrees clockwise (only
      really works well with block-fonts, like "banner3"
    * you can now use <random> for the font or cowfile, and it will pick
      one for you.

9.0: Tue Jun  1 18:28:08 2004
    * add -mirc flag to force mirc colors from commandline
    * added zb's new cowfiles.  best ever.

8.9: Wed May  5 09:20:40 2004
    * Fix utf8 bug for reals

8.8: Wed Apr 28 22:52:36 2004
    * add -bash flag to do escaping (see POD for details)
    * bugfix: don't try utf8 in perl < 5.8
    * begin POD, first item: bash escaping

8.7: Mon Apr 26 12:56:43 2004
    * 8.7 changes submitted by j0no:
    * AOL Instant Messenger support.
    * phpBB bbcode support
    * orkut support
    * Fixed mIRC code hack. (\x0F instead of \x03\26\26)
    * & -> &amp; in HTML and AIM
    * -cmd flag to run any irssi command! Use quotes for stuff like "kick
      zb"
    * -utf8 flag for cat
    * Bugfix: Rainbow Offset saved across figs.
    * -8 Canada colour map
    * Added check to skip the last mile if there's nothing to process.
    * Added a -mirc flag to force mIRC codes
    * clean_length() repaired
    * other fixes:
    * verified all filters handle undef properly

8.6: Tue Apr 13 15:17:51 2004
    * add flags for cli: -stdin will use data from STDIN
    * add flags for cli: -cat, -exec to hook irssi functionality
    * stop processing if we don't have text after first pass

8.5: Tue Apr 13 13:00:02 2004
    * remove legacy open3() and use IPC::Open3 -- works in 5.8 now
    * support for specifying color in -4 .. for example -4red
    * added some more spookwords
    * added the -unused command

8.3: Sat Jan 17 23:03:50 2004
    * added -ansi flag to force ANSI output in cases where it is not
      detected properly and tries to use irc colors
    * default html colors has white foreground
    * no warnings if $COWPATH isn't set
    * no more Text::Wrap warnings

8.2: Tue Jan 13 19:01:09 2004
    * If not Irssi, rainbow keepstate is down with a dotfile
    * Bugfix: HTML < and > entity escaping fixed
    * Bugfix: IRC output doesn't use ANSI resets (breaks xchat! thanks
      kin)
    * Optimization: Don't change HTML bgcolor unless it differs from the
      default black.. saves lots of space
    * Make Term::ANSIColor optional (ansi is disabled if no lib)
    * Silenced the throttling message, what's the point of THAT
    * Cleaned up the COWPATH code a bit.. better detection
    * html output has div container to set background to black, for
      pasting chunks of html, e.g. to livejournal.
    * Abstract prefix code to its own filter, for code consistency

8.1: Tue Jan 13 00:53:55 2004

color/color.py
#!/usr/bin/env python

import sys
import re
import codecs
import optparse

class ColorMapError(Exception):

    pass


class ColorMap(object):

    """Class to manage an ANSI/mIRC color-mapped text string"""

    # mapping by color name
    colnames = ['black',    # 0
                'red',      # 1
                'green',    # 2
                'yellow',   # 3
                'blue',     # 4
                'magenta',  # 5
                'cyan',     # 6
                'white']    # 7

    # mapping by color code
    codes = ['d', 'r', 'g', 'y', 'b', 'm', 'c', 'w']

    # structural data for various color schemes
    schemes = {'mirc': {'color_re': re.compile('(\x03[0-9,]+\x16*|\x0f)'),
                        'parse_re': re.compile('\x03([0-9]+)?(?:,([0-9]+))?'),
                        # map mirc colors to ansi intense/color codes
                        'map': [(1, 7),    # 0
                                (0, 0),    # 1
                                (0, 4),    # 2
                                (0, 2),    # 3
                                (1, 1),    # 4
                                (0, 1),    # 5
                                (0, 5),    # 6
                                (0, 3),    # 7
                                (1, 3),    # 8
                                (1, 2),    # 9
                                (0, 6),    # 10
                                (1, 6),    # 11
                                (1, 4),    # 12
                                (1, 5),    # 13
                                (1, 0),    # 14
                                (0, 7)]},  # 15
               'ansi': {'color_re': re.compile('(\x1b\[[0-9;]+m)'),
                        'parse_re': re.compile('\x1b\[([0-9;]+)')},
               # dummy entry to handle uncolored text
               'plain': {'color_re': re.compile('(\xff)'),
                         'parse_re': re.compile('(\xff)')}}

    # the default color scheme (white on black)
    reset = (7,  # foreground color
             0,  # intensity on
             0,  # background color
             1)  # default flag indicates color has not changed

    def __init__(self, data=None, scheme=None, encoding=None):
        """Instantiate a color-mapped text string"""

        # try to determine encoding of source data
        if not encoding and hasattr(data, 'encoding'):
            encoding = data.encoding
        try:
            encoding = codecs.lookup(encoding).name
        except:
            encoding = sys.getdefaultencoding()

        # file-like objects return lines with newlines intact,
        # other iterables return stripped lines
        if hasattr(data, 'read'):
            data = ''.join(data)
        elif hasattr(data, '__iter__'):
            data = '\n'.join(data)

        # make sure data is unicode
        if data is None:
            data = u''
        elif isinstance(data, str):
            data = data.decode(encoding, 'replace')
        if not isinstance(data, unicode):
            raise ColorMapError('unknown data type: %s' % type(data))

        # verify scheme is valid
        if scheme is None:
            scheme = self.detect(data)
        if scheme not in self.schemes:
            raise ColorMapError('unknown scheme: %s' % scheme)

        # store decoded version
        self.plain, self.colmap = self.decode(data, scheme)
        self.scheme = scheme
        self.encoding = encoding

    def render(self, scheme=None):
        if scheme is None:
            scheme = self.scheme
        data = self.encode(self.plain, self.colmap, scheme)
        data = '\n'.join(data).encode(self.encoding, 'replace')
        return data

    @classmethod
    def detect(cls, data):
        """Try to guess color scheme"""
        seen = {}
        for scheme in cls.schemes:
            seen[scheme] = len(cls.schemes[scheme]['color_re'].findall(data))
        if sum(seen.values()):
            return sorted(seen.iteritems(),
                          key=lambda item: item[1],
                          reverse=True)[0][0]
        return 'plain'

    @classmethod
    def decode(cls, data, scheme):
        """Returns decoded plain text and its colormap from text"""

        # sanity check inputs
        if not isinstance(data, unicode):
            raise ColorMapError('decode requires unicode data')
        if scheme not in cls.schemes:
            raise ColorMapError('unknown scheme: %s' % scheme)

        # decode
        plain = []
        colmap = []
        attrs = cls.schemes[scheme]
        for line in data.splitlines():
            plain_line = []
            colmap_line = []
            fg, intense, bg, default = cls.reset
            for part in attrs['color_re'].split(line):
                if not part:
                    continue
                if attrs['color_re'].match(part):
                    if part == '\x0f':
                        fg, intense, bg, default = cls.reset
                    else:
                        part = attrs['parse_re'].search(part).groups()
                        default = 0
                        if scheme == 'mirc':
                            new_fg, new_bg = part
                            if new_fg is not None:
                                intense, fg = attrs['map'][int(new_fg)]
                            if new_bg is not None:
                                bg = attrs['map'][int(new_bg)][1]
                        elif scheme == 'ansi':
                            for part in [int(i) for i in part[0].split(';')]:
                                if part == 0:
                                    fg, intense, bg, default = cls.reset
                                elif part == 1:
                                    intense = 1
                                elif part in (2, 22):
                                    intense = 0
                                elif part == 39:
                                    fg = cls.reset[0]
                                elif part == 49:
                                    bg = cls.reset[2]
                                elif part >= 30:
                                    part -= 30
                                    if part >= 10:
                                        bg = part - 10
                                    else:
                                        fg = part
                else:
                    plain_line.append(part)
                    col = cls.pack(fg, intense, bg, default)
                    colmap_line.append(col * len(part))
            plain.append(''.join(plain_line))
            colmap.append(''.join(colmap_line))
        return plain, colmap

    @classmethod
    def encode(cls, lines, colmap, scheme):
        """Encode in color"""

        # sanity check inputs
        if scheme not in cls.schemes:
            raise ColorMapError('unknown scheme: %s' % scheme)

        # encode
        attrs = cls.schemes[scheme]
        output = []
        for i, line in enumerate(lines):
            last = list(cls.reset)  # the last color we painted
            output_line = []        # one fully rendered line
            for j, ch in enumerate(line):

                # first get the color mapped to this character
                try:
                    col = list(cls.unpack(colmap[i][j]))
                except IndexError:
                    col = last

                # what's changed?
                fg_changed = col[0] != last[0]
                intense_changed = col[1] != last[1]
                bg_changed = col[2] != last[2]

                # the conditions under which we will need to repaint
                if (bg_changed or
                    (ch != ' ' and (fg_changed or intense_changed))):

                    outcol = None
                    if scheme == 'mirc':
                        if col[3]:
                            outcol = '\x0f'
                        else:
                            codes = []
                            if fg_changed or intense_changed:
                                newcol = attrs['map'].index((col[1], col[0]))
                                codes.append(str(newcol))
                            if bg_changed:
                                if not codes:
                                    codes.append('')
                                newcol = attrs['map'].index((0, col[2]))
                                codes.append(str(newcol))
                            outcol = '\x03%s' % ','.join(codes)
                            if (j + 1) < len(line) and line[j + 1].isdigit():
                                outcol += '\x16\x16'
                    elif scheme == 'ansi':
                        codes = []
                        if col[3]:
                            codes.append(0)
                        else:
                            if fg_changed:
                                codes.append(col[0] + 30)
                            if bg_changed:
                                codes.append(col[2] + 40)
                            if intense_changed:
                                if col[1]:
                                    codes.append(1)
                                else:
                                    codes.append(22)
                        outcol = '\x1b[%sm' % ';'.join(map(str, codes))
                    if outcol:
                        last = col
                        output_line.append(outcol)

                output_line.append(ch)
            output.append(''.join(output_line))
        return output

    @staticmethod
    def pack(fg, intense=0, bg=0, default=1):
        """Pack color -> char"""
        return chr(fg + (intense << 3) + (bg << 4) + (default << 7))

    @staticmethod
    def unpack(char):
        """Unpack color -> (fg, intense, bg, default)"""
        char = ord(char)
        return (char & 7), (char & 8) >> 3, (char & 112) >> 4, (char & 128) >> 7

    @classmethod
    def pack_by_name(cls, name):
        """Return the packed color value by name"""

        # XXX this whole function should be greatly simplified..
        # but at least it works, will be useful somewhere i think.
        name = name.lower()
        name = name.replace('dark', '')
        name = name.replace('light', 'bright')
        name = name.replace('gray', 'grey')
        name = name.replace('orange', 'bright red')
        name = name.replace('purple', 'magenta')
        name = name.split()
        if 'on' not in name:
            name += ['on', 'black']
        on = name.index('on')

        # grey -> bright black
        # bright grey -> white
        # white -> bright white
        def fixgrey(words):
            if 'grey' in words:
                if 'bright' in words:
                    return ['white']
                else:
                    return ['bright', 'black']
            elif 'white' in words:
                return ['bright', 'white']
            else:
                return words

        fg, intense, bg, default = cls.reset
        for word in fixgrey(name[:on]):
            if word == 'bright':
                intense = 1
                default = 0
            elif word in cls.colnames:
                fg = cls.colnames.index(word)
                default = 0
        for word in fixgrey(name[on + 1:]):
            if word in cls.colnames:
                bg = cls.colnames.index(word)
                default = 0

        char = cls.pack(fg, intense, bg, default)
        return char

    def __iter__(self):
        for line in self.plain:
            yield line

    def __str__(self):
        return '\n'.join(self).encode(self.encoding, 'replace')

    def __repr__(self):
        return '<%s object at 0x%x: lines=%d, scheme=%s, encoding=%s>' % (
                self.__class__.__name__, id(self), len(self.plain),
                self.scheme, self.encoding)


#
# so.. how to continue this project?  it might be in my best interest
# to try and wedge what i have so far onto insub.py to see if this
# will even work, before i refine it much futher..

##########################################################################
### The following functions are just some hacks to mess around w/ maps ###
##########################################################################

"""
	rainbow	=> 'rrooyyYYGGggccCCBBbbmmMM',	# -1
	usa	=> 'oowwBB',			# -2
	blue	=> 'bB',			# -4blue
	green	=> 'gG',			# -4green
	purple	=> 'mM',			# -4purple
	grey	=> '12',			# -4grey (-5)
	yellow	=> 'yY',			# -4yellow
	red	=> 'or',			# -4red
	scale	=> 'ww22CC11CC22',		# -6
	xmas	=> 'og',			# -7
	canada	=> 'ooww',			# -8
"""

# this brutally expands a map by some factor. with no anti-aliasing,
# or sub-pixel hinting (loolol), you probably want to avoid this ever being
# called. in practice this means having source maps that are larger than
# the ascii text they are painting.

def expand(lines, factor):
    new = []
    for line in lines:
        new += [''.join(ch * factor for ch in line)] * factor
    return new

# this shrinks a map to a specific size by dropping some factor of pixels
# along a line, and skipping some lines.  this has a few problems.  namely,
# removing a whole line from a smaller bitmap can completely change what
# it's supposed to be.  second, image maps created from an image with 1:1
# pixel aspect ratio become all stretchy.  this function should be able to
# remove lines at a higher frequency so as to stretch the image back into
# shape..
def shrink(lines, width):
    longest = len(max(lines, key=len))
    if isinstance(width, float) and width < 1:
        width = int(longest * width)
    scale = int(float(longest) / width * 100)
    hscale = int((float(longest) * 1) / width * 100)
    scaled = []
    for i in xrange(len(lines) * 100):
        line = lines[i / 100]
        if not i % scale:
            scaled_line = []
            for j in xrange(len(line) * 100):
                ch = line[j / 100]
                if not j % hscale:
                    scaled_line.append(ch)
            scaled.append(''.join(scaled_line))
    return scaled


# this function takes two bitmaps, refits the latter to fit the former
# by any means necessary (shrinking/expanding), and returns it.  it can
# be pretty damn slow, so we might need to rethink this.
def fit(map, width):
    factor = 1
    while True:
        if factor > 1:
            new = expand(map, factor)
        else:
            new = map
        if len(new[0]) >= width:
            map = new
            break
        factor += 1
    if len(map[0]) > width:
        map = shrink(map, width)
    return map


# this takes a bitmap in the external format (using one-letter codes)
# and puts it into a format that ColorMap can understand.  This could
# probably be integrated into colormap
def repack(map):
    new = []
    for line in map:
        new_line = []
        for ch in line:
            if ch.isupper():
                intense = 1
                ch = ch.lower()
            else:
                intense = 0
            fg = ColorMap.codes.index(ch)
            col = ColorMap.pack(fg, intense, default=0)
            new_line.append(col)
        new.append(''.join(new_line))
    return new


def main():
    parser = optparse.OptionParser()
    parser.add_option('-m', '--mirc', dest='scheme', default='ansi',
                      action='store_const', const='mirc')
    parser.add_option('-w', '--width', default=67, type='int')
    opts, args = parser.parse_args()

    for path in args:
        with open(path, 'rb') as fp:
            colmap = fp.read()
        colmap = fit(colmap.splitlines(), opts.width)
        plain = ['*' * opts.width] * len(colmap)
        map = ColorMap(plain, opts.scheme)
        map.colmap = repack(colmap)
        print map.render()
    return 0


if __name__ == '__main__':
    try:
        import psyco
        psyco.cannotcompile(re.compile)
        psyco.full()
    except ImportError:
        pass
    sys.exit(main())
